Single-path symex checker [blocks: #3976]

peterschrammel opened the pull request:
Implements the --paths variant of using goto-symex.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Factor out output_coverage_report into bmc_util

Reduces code duplication across goto checkers.

peterschrammel committed:
Add single-path symex-only checker

Replaces the path_explorert variant of bmct.
The current implementation of this is overly complicated
because path_explorert distinguishes the first run of goto-symex
from the following ones. To drastically simplify the code
goto-symex/symex_bmc needs to be refactored so that all
runs can be handled uniformly (e.g. push entry point into work list
and then 'resume' from there).
See also single_path_symex_checkert.

peterschrammel committed:
Use goto verifier for --paths with show-vcc and -program-only in CBMC

We can now use single_path_symex_only_checkert instead of bmct.

peterschrammel committed:
Add single-path symex checker

Extends single_path_symex_only_checkert with
calling the solver and returning traces.

peterschrammel committed:
Use goto verifier for --paths in CBMC

We can now use single_path_symex_checkert instead of bmct.

peterschrammel committed:
Add SSA validation

As introduced into bmct in eb0b86be.

peterschrammel committed:
Factor out perform_symex into bmc_util

Better provide it as a utility than hiding it inside a class.
Reduces code duplication across goto checkers.

peterschrammel committed:
Add perform_symex for resuming from state

Copied from bmct. Going to be used by single_path_symex_checkert.

peterschrammel committed:
Factor out goto_symex_property_decider

Provides functionality close to solvers/prop/cover_goals but without
callbacks. It allows us to iteratively get all counterexamples
for a set of properties on a fixed equation.
This is used as a component by multi-path and single-path
symex checkers.

peterschrammel committed:
Set final status of UNKNOWN properties

Utility function to set all UNKNOWN properties to PASS
when the verification algorithm has finished.
Corresponds to the current behaviour of CBMC replying
'successful' when the given bounds have been reached.

peterschrammel committed:
Single-path goto checker for Java requires preprocessing

JBMC requires further setup of symex_bmct.

peterschrammel committed:
Use goto verifier for --paths in JBMC

We can now use single_path_symex_checkert instead of bmct.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0420c96f4c3c483e0f62e1bed5f79726f148fffc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98863639
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented on src/goto-checker/bmc_util.h:
```diff
@@ -69,6 +69,12 @@ void slice(
   const optionst &,
   ui_message_handlert &);
 
+void output_coverage_report(
```
Documentation would be appreciated. I think the `cov_out` parameter is the one needing it the most, notably the aspect that an empty string will mean that nothing happens. How about:
```
/// Output a coverage report as generated by \ref symex_coveraget if \p cov_out is non-empty.
/// \param cov_out: file to write the report to; no report is generated if this is empty
/// \param goto_model: goto model to build the coverage report for
/// \param symex: symbolic execution run to report coverage for
/// \param ui_message_handler: status/warning message handler
```

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f94473badd9eb23d35ae76594ebb9ce6d7fcccbe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99520817

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
+{
+  // add a partial ordering, if required
+  // We won't be able to decide any properties by adding this,
+  // but we'd like to see the entire SSA.
```
Does this comment still make sense? I recall requesting it, but now that this is more generic I don't think it's appropriate anymore.

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
+    return goto_model.get_goto_function(id);
+  };
+
+  // perform symbolic execution
```
Comments don't always add value: I think it's pretty clear what the next line does, and this just makes it sound like something surprising would happen?

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
+    [&goto_model](
+      const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Should we maybe tweak the `goto_symext` interface a bit? Looking at `get_function_from_goto_functions` (`symex_main.cpp`) we now define this helper function multiple times.

tautschnig commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -336,3 +336,25 @@ void perform_symex(
   postprocess_equation(symex, equation, options, ns, ui_message_handler);
 }
+
+void perform_symex(
```
I'm confused. It seems two commits now introduce the same. They actually do, the lines above already contain the same code.

peterschrammel replied:
Yes, that's exactly the problem.  The goto-symex interface is broken in this respect. We should have a function that returns a `path_storaget::patht` for the initial state at the entry point; and then we can `resume` from that... and everything becomes nice and uniform.

tautschnig replied:
That should be fixable?

peterschrammel replied:
Yes, but I'll go ahead with the current version for now and clean up the goto-symex interface later this week. https://github.com/diffblue/cbmc/issues/4071

tautschnig commented on src/goto-checker/bmc_util.h:
```diff
@@ -79,6 +80,16 @@ void perform_symex(
   ui_message_handlert &);
 
+/// Perform symbolic execution starting from \p resume.
+void perform_symex(
```
Declaration is also added twice.

tautschnig commented on src/goto-checker/bmc_util.h:
```diff
@@ -117,6 +117,16 @@ void update_status_of_not_checked_properties(
   std::unordered_set<irep_idt> &updated_properties);
 
+/// Sets the property status of UNKNOWN properties to PASS.
+/// \param [inout] properties: The status is updated in this data structure
```
The coding standard uses `in,out`

tautschnig commented on src/goto-checker/goto_symex_property_decider.cpp:
```diff
@@ -0,0 +1,162 @@
+    ui_message_handler,
+    ui_message_handler.get_ui() == ui_message_handlert::uit::XML_UI);
+  solver = solvers.get_solver();
+  solver->prop_conv().set_message_handler(ui_message_handler);
```
That should really be the `solver_factoryt`'s job. Invoking `get_solver()` ought to make sure the message handler is set up, the factory does have all the necessary info.

tautschnig commented on src/goto-checker/goto_symex_property_decider.cpp:
```diff
@@ -0,0 +1,162 @@
+
+exprt goto_symex_property_decidert::goalt::as_expr() const
+{
+  std::vector<exprt> tmp;
```
`exprt::operandst conjuncts;`

tautschnig commented on src/goto-checker/Makefile:
```diff
@@ -7,6 +7,7 @@ SRC = bmc_util.cpp \
       goto_verifier.cpp \
       multi_path_symex_checker.cpp \
       multi_path_symex_only_checker.cpp \
+      single_path_symex_only_checker.cpp \
```
Lexicographic sorting, please. Applies again in a later commit (introducing `single_path_symex_checker.cpp`).

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -559,13 +560,22 @@ int cbmc_parse_optionst::doit()
+        options, ui_message_handler, goto_model);
+    }
+
+    (void)(*verifier)();
```
Seems like an unnecessary dance, can't we just run each of them in the branches above?

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -581,13 +582,22 @@ int cbmc_parse_optionst::doit()
+          options, ui_message_handler, goto_model);
     }
+
+    (void)(*verifier)();
```
Again, seems unnecessary. Just call each of them in the branches above.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -559,46 +561,83 @@ int jbmc_parse_optionst::doit()
-        return CPROVER_EXIT_SUCCESS;
       }
+
+      (void)(*verifier)();
```
Just call each of them in the branches above.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -559,46 +561,83 @@ int jbmc_parse_optionst::doit()
+          options, ui_message_handler, goto_model);
+      }
+
+      (void)(*verifier)();
```
Just call each of them.

peterschrammel committed:
Use in,out for in/out params in doxygen

peterschrammel committed:
Set message handler in solver factory

Allows us to remove set_message_handler() calls on solvers
in other parts of the code.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1f9f4756d8eb7f6fba819ba2d3913fca3517a61e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99702092
